crates/cargo-contract/src/cmd/extrinsics/events.rs

```diff
@@ -19,100 +19,175 @@ use super::{
     DefaultConfig,
 };
 use crate::{
-    maybe_println,
     Verbosity,
     DEFAULT_KEY_COL_WIDTH,
 };
 use colored::Colorize as _;
 use transcode::{
     ContractMessageTranscoder,
     TranscoderBuilder,
+    Value,
 };
 
-use anyhow::Result;
+use anyhow::{
+    Ok,
+    Result,
+};
+use std::fmt::Write;
 use subxt::{
     self,
     events::StaticEvent,
     tx::TxEvents,
 };
 
-pub fn display_events(
-    result: &TxEvents<DefaultConfig>,
-    transcoder: &ContractMessageTranscoder,
-    subxt_metadata: &subxt::Metadata,
-    verbosity: &Verbosity,
-) -> Result<()> {
-    if matches!(verbosity, Verbosity::Quiet) {
-        return Ok(())
+/// Field that represent data of the event from contract call
+#[derive(serde::Serialize)]
+pub struct Field {
+    /// name of a field
+    pub name: String,
+    /// value of a field
+    pub value: Value,
+}
+
+impl Field {
+    pub fn new(name: String, value: Value) -> Self {
+        Field { name, value }
     }
+}
 
-    println!();
+/// Events produced from calling a contract
+#[derive(serde::Serialize)]
+pub struct Event {
+    /// name of a pallet
+    pub pallet: String,
+    /// name of the event
+    pub name: String,
+    /// data associated with the event
+    pub fields: Vec<Field>,
+}
 
-    if matches!(verbosity, Verbosity::Verbose) {
-        println!("VERBOSE")
-    }
+/// Result of the contract call
+#[derive(Default, serde::Serialize)]
+pub struct CallResult {
+    /// Instantiated contract hash
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub contract: Option<String>,
+    /// Instantiated code hash
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub code_hash: Option<String>,
```

Comment:
A more strongly typed way to achieve this without having to set these optional fields, would be to defined a common "inner" struct:

```
#[derive(serde::Serialize)]
struct ExtrinsicResult {
  events; Vec<Event>
  ..
}
```
Then different top level structs with the varying details, flattening the inner struct:
```
#[derive(serde::Serialize)]
struct InstantiateWithCodeResult {
  contract: String,
  #[serde(flatten)]
  details: ExtrinsicResult,
}

#[derive(serde::Serialize)]
struct InstantiateResult {
  code_hash: String,
  #[serde(flatten)]
  details: ExtrinsicResult,
}
```

Reply:
`code_hash` is still an optional field because of this logic in `instantiate_with_code()`
```rust
// The CodeStored event is only raised if the contract has not already been uploaded.
let code_hash = result
    .find_first::<api::contracts::events::CodeStored>()?
    .map(|code_stored| code_stored.code_hash);
// <--snippet--> 
call_result.contract = Some(instantiated.contract.to_ss58check());
```

So, `contract` is always present in the struct, but `code_hash` field may be `None`. Is there any point defining common inner struct?

Reply:
You're right it is probably not necessary, the main issue is that we should avoid the pattern of having a `mut CallResult` and setting the values of the public fields.

The values should be set in the constructor

Reply:
Ok I have made a refactoring PR to show how I might do it, along with some other stuff https://github.com/paritytech/cargo-contract/pull/730

Feel free to reject the PR or change it, it is just some ideas about refactorings I would do.